### DISCUSSION TITLE
[v5] hasOne preload return error undefined in null id

### DISCUSSION THREAD

**Discussion by arthurcvm** [adonisjs/core#2607] [Help] [answered]
In my system I make a CRUD operation and after make a refresh followed by a preload in all related models, this was working normally for belongsTo, but in hasOne relationship return an error when the id is null, look the code:

```ts
await sector.refresh()
await sector.load((loader) => {
    loader.preload('supervisor').preload('coordinator')
      .preload('manager').preload('department')
})
```
The exception returned:

`Exception: Cannot preload "supervisor", value of "Sector.supervisor_id" is undefined`

But in belongsTo relationship this code work.
What's the problem?

**Comment by arthurcvm** [comment, marked as answer]
I understood the [docs](https://docs.adonisjs.com/guides/models/relationships), in Sector I set belongsTo and in Supervisor I set hasOne, this way it works.

**Reply by d4n1b** [reply]
I had a hard time understanding this too. It's safe for `null` values when using the combination you mentioned.

In my case I have `Activities` & `Clients`.

This does **NOT** work when `client_id` is null ❌ :

```
export default class Activity extends BaseModel {
  @hasOne(() => Client)
  public client: HasOne<typeof Client>;
```


This **DOES** work when `client_id` is null  ✅ :
```
export default class Activity extends BaseModel {
  @belongsTo(() => Client)
  public client: BelongsTo<typeof Client>;
```

**Reply by thetutlage** [reply]
For that you need to understand when to use `belongsTo` and when to use `hasOne`. In your case it should be `belongsTo`. A simple way to remember is. 

A model that has the foreign key makes a belongs to relationship with the other model

**Reply by d4n1b** [reply]
Thanks @thetutlage - It will be definitely easier to remember now :smile::+1:
